File: 257. Binary Tree Paths.py
# https://leetcode.com/problems/binary-tree-paths/

# Given the root of a binary tree, return all root-to-leaf paths in any order.

# A leaf is a node with no children.

################################################################################

# dfs(node)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: TreeNode) -> List[str]:
        if not root: return []
        
        self.ans = []
        self.holder = [str(root.val)]
        self.dfs(root)
        return self.ans
    
    def dfs(self, node):
        if not node.left and not node.right:
            self.ans.append("->".join(self.holder))
            return
        
        if node.left:
            self.holder.append(str(node.left.val))
            self.dfs(node.left)
            self.holder.pop()
        if node.right:
            self.holder.append(str(node.right.val))
            self.dfs(node.right)
            self.holder.pop()

File: 257. Binary Tree Paths.cpp
// https://leetcode.com/problems/binary-tree-paths/

// Given the root of a binary tree, return all root-to-leaf paths in any order.

// A leaf is a node with no children.

////////////////////////////////////////////////////////////////////////////////

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        holder.push_back(root->val);
        dfs(root);       
        return ans;
    }
private:
    vector<int> holder;
    vector<string> ans;
    void dfs(TreeNode* node) {
        if (!node->left and !node->right) {
            ans.push_back(toPath(holder)); return;
        }
        if (node->left) {
            holder.push_back(node->left->val);
            dfs(node->left);
            holder.pop_back();
        }
        if (node->right) {
            holder.push_back(node->right->val);
            dfs(node->right);
            holder.pop_back();
        }
        return;
    }
    string toPath(const vector<int>& holder) {
        string path = to_string(holder[0]);
        for (int i = 1; i < holder.size(); ++i) {
            path += "->" + to_string(holder[i]);
        }
        return path;
    }
};
